src/lib/container/smartlist.c

```diff
@@ -79,6 +79,25 @@ smartlist_string_remove(smartlist_t *sl, const char *element)
   }
 }
 
+/** Return <b>sl</b> of size <b>size</b>, removing and freing all the extra
+ * elements. Preserve order. */
+void
+smartlist_truncate(smartlist_t *sl, const size_t size)
+{
+  int i, diff;
+  /* Difference between the number of elements in sl and the maximum size */
+  diff = sl->num_used - size;
+  /* Remove elements at position size, difference times, the next element
+   * index will be always size. */
+  for (i=0; i < diff; i++) {
+    log_debug(LD_DIRSERV, "sl->list[size] %s\n", (char *)sl->list[size]);
+    log_debug(LD_DIRSERV, "sl len %d\n", smartlist_len(sl));
+    smartlist_del(sl, size);
```

Comment:
Use smartlist_del_keeporder() to avoid reordering the bandwidth headers.
Use smartlist_del_keeporder(sl, i) to delete the correct item.
Use tor_free(smartlist_get(sl, i)) to free the item before deleting it.

Also, you should probably delete from the end of the smartlist, rather than the start.

Reply:
Actually, I think we can delete this entire commit.
See my comments on the next commit.